### PR TITLE
Http2 chapter

### PR DESCRIPTION
**Pull request by bazzadp**
Closes https://github.com/HTTPArchive/almanac.httparchive.org/issues/173

@bagder, @dotjs , @rmarx feel free to add further feedback as reviews on this PR.

### CODE REVIEW

File: src/content/en/2019/http2.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+This whistle-stop tour of HTTP/2 gives the main history and concepts of the newish protocol. As should be apparent from this explanation, the main benefit of HTTP/2 is to address performance limitations of the HTTP/1.1 protocol. There were also security improvements as well - perhaps most importantly in being to address performance issues of using HTTPS since HTTP/2, even over HTTPS, is [often much than plain HTTP](https://www.httpvshttps.com/). Other than the web browser packing the HTTP messages into the new binary format, and the web server unpacking it at the other side, the core basics of HTTP itself stayed roughly the same. This means web applications do not need to make any changes to support HTTP/2 as the browser and server take care of this. Turning it on should be a free performance boost and because of this adoption should be relatively easy. Of course, there are ways web developers can optimize for HTTP/2 to take full advantage of how it differs.
```
Comment (rmarx) [inline src/content/en/2019/http2.md]: often much than plain HTTP > often much faster than plain HTTP?

File: src/content/en/2019/http2.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+|               |   4.56% |  5.13% |  4.87% |
+| LiteSpeed     |   4.11% |  4.97% |  4.57% |
+| GSE           |   2.16% |  3.73% |  3.01% |
+| Microsoft-IIS |   3.09% |  2.66% |  2.86% |
+| openresty     |   2.15% |  2.01% |  2.07% |
+| ...           |   ...   |  ...   |  ...   |
+
+**Figure 5 - Servers used for HTTP/2**
+
+Nginx provides package repos that allow ease of installing or upgrading to a the latest version, so it is no surprise to see it leading the way here. Cloudflare is the [most popular CDN](../cdn/) and enables HTTP/2 by default so again it is also not surprising to see this as a large percentage of HTTP/2 sites. Incidently, Cloudflare uses [a heavily customised version of nginx as their web server](https://blog.cloudflare.com/nginx-structural-enhancements-for-http-2-performance/). After this we see Apache at around 20% of usage, followed by some servers who choose to hide what they are and then the smaller players (LiteSpeed, IIS, Google Servlet Engine and openresty - which is nginx based).
```
Comment (rmarx) [inline src/content/en/2019/http2.md]: to a the latest version > to the latest version

File: src/content/en/2019/http2.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+| Desktop |  7.86               | 162.38        |
+| Mobile  |  6.35               | 122.78        |
+
+**Figure 11 - How much is pushed when it is used**
+
+This is a concern as previous advice has been to be conservative with push and to ["push just enough resources to fill idle network time, and no more"](https://docs.google.com/document/d/1K0NykTXBbbbTlv60t5MyJvXjqKGsCVNYHyLEXIxYMv0/edit). The above statistics suggest many resources, of a significant combined size are pushed. Looking at what is pushed we see the data in Figure 12:
+
+![](https://www.tunetheweb.com/assets/images/whatpushisusedfor.png)
+
+**Figure 12 - What asset types is pushed used for?**
```
Comment (rmarx) [inline src/content/en/2019/http2.md]: What asset types is pushed used for > What asset types is push used for

File: src/content/en/2019/http2.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
..., that the browser will not see until the CSS for example has been requested, downloaded and parsed. If these are now pushed based on that header, there was a concern that reusing this may result in a lot of unintended pushes. However, the relative low usage of fonts and images may mean that risk is not being seen as much as was feared. `<link rel="preload" ... >` tags are often used in the HTML rather than HTTP link headers and the meta tags are not a signal to push. Statistics in the [resource hint](../resourcehints/) chapter show that less tha 1% of sites use the preload HTTP link header, and about the same amount use preconnection which has no meaning in HTTP/2 so this would suggest this is not so much of an issue. Though there are a number of fonts and other assets being pushed, which may be a signal of this. As a counter argument to those complaints, if an asset is important enough to preload, then it could be argued these assets should be pushed if possible as browsers treat a preload hints as very high priority requests anyway. Any performance concern is therefore (again arguably) at the overuse of preload, rather than the resulting HTTP/2 push that happens because of this.
```
Comment (rmarx) [inline src/content/en/2019/http2.md]: show that less tha > show that less than

**Comment by rmarx** [comment]
Changes look good to me :) Just a few typo's I found and marked above. Looking forward to seeing this in the Almanac!

**Comment by bazzadp** [comment]
Thanks corrected those of you could run another pair of eyes over them to make sure I did them correctly @rmarx?

File: src/content/en/2019/http2.md [hunk 5]
```diff
@@ -87,7 +87,6 @@ We have shown that browser support is strong, and there is a safe road to adopti
 
 | Server        | Desktop | Mobile | Both   |
 | ------------- | ------- | -------| ------ |
-| nginx         |  34.04% | 32.48% | 33.19% |
```
Comment (rmarx) [inline src/content/en/2019/http2.md]: it seems like this change was un-intentional?

**Comment by rmarx** [comment]
Seems like you accidentally removed the nginx-line from one of the tables (indicated in a comment on the PR). Other than that, looks good.

**Comment by bazzadp** [comment]
And that's why a second pair of eyes is good! Thanks @rmarx fixed now.

**Comment by bazzadp** [comment]
Thanks @rviscomi ! Three questions for you:

1. I've referenced other chapters in this chapter but what's the best to hyper link them? At the moment I've gone with the below but not show of the exact folder structure or filename they will have (`/resourcehinsts/` or `/resourcehints.html` or something else?):

`One of the most popular uses of the `preload` [resource hint](../resourcehints/) is to inform...`

2. I currently have 4 images in this document based on the stats. At the moment they are just screenshots hosted on my site as I'm presuming they will be replaced later with nicer graphics. Is that OK or would you prefer I add even these temporary images to the chapter's graphics folder? Guess that might be useful in case we run out of time to replace them with proper images before launch. Anyway let me know how best to represent these.

3. I've quite a few tables in this chapter for the data. I've just used markdown to draw them. Is that the preferred method or should I reference the stats they represent somehow to allow the data people to draw them better as fancy tables?

**Comment by rviscomi** [comment]
@bazzadp great questions.

1. Leave a placeholder link for now, eg `[resource hints](#)` and we'll replace it during editing.
2. You can add images for your chapter to the [src/static/images/2019/20_HTTP_2](https://github.com/HTTPArchive/almanac.httparchive.org/tree/master/src/static/images/2019/20_HTTP_2) directory
3. Markdown tables are fine. We'll have table CSS to make them look pretty. Is there a spreadsheet with all of the chapter's results somewhere? If so we can generate data viz in that sheet and embed it on the page (synced with sheet and interactive), also please add me as an editor (cc @paulcalvano).

**Comment by bazzadp** [comment]
Yes there is a data sheet here: https://docs.google.com/spreadsheets/d/1z1gdS3YVpe8J9K3g2UdrtdSPhRywVQRBz5kgBeqCnbw/edit?usp=sharing.

Not sure how to embed data from there in markdown to be honest but only need one image from there (Figure 12). The rest (Figures 1, 8 and 9) are from https://httparchive.org/reports/state-of-the-web, but couldnt get them to display as nicely as those so just took screenshots of them. I did copy the data into the above sheet.

**Comment by bazzadp** [comment]
@dotjs did you get a chance to review this anymore? Can you let me know if you intend to send further comments and, if not, will then merge this pull request (or @rviscomi should you click the merge button rather than me?).

**Comment by rviscomi** [comment]
I've pinged @dotjs on Slack 12 hours ago and gotten no reply, so let's go ahead and marge the PR. Thanks @bazzadp for your hard work on this, it came out excellent!

**Pull request merged**